File: backend/services/stigmer-server/pkg/domain/agentexecution/filereview/gate.go
package filereview

import (
	agentexecutionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/agentexecution/v1"
)

// The HITL resume gate is UNIFIED across two lifecycles that can both block a
// single turn: tool approvals (pending_approvals) and file review
// (file_change_sets awaiting a human decision). A turn is unblocked only when
// BOTH are clear. These helpers are the one place that truth is computed, so the
// workflow's wake condition, SubmitApproval's signal, and SubmitFileDecision's
// signal can never disagree about whether the turn may resume.

// CountAwaitingReview returns how many change sets are still waiting on a human
// decision (status AWAITING_REVIEW). A set that is CAPTURING is mid-turn (the
// runner has not returned yet) and a DECIDED/RECONCILED/FAILED set no longer
// blocks the gate, so neither counts.
func CountAwaitingReview(changeSets []*agentexecutionv1.FileChangeSet) int {
	n := 0
	for _, cs := range changeSets {
		if cs.GetStatus() == agentexecutionv1.FileChangeSetStatus_FILE_CHANGE_SET_STATUS_AWAITING_REVIEW {
			n++
		}
	}
	return n
}

// NoChangeSetAwaitingReview reports whether the file-review sub-gate is clear:
// no change set is waiting on a human decision.
func NoChangeSetAwaitingReview(status *agentexecutionv1.AgentExecutionStatus) bool {
	return CountAwaitingReview(status.GetFileChangeSets()) == 0
}

// UnresolvedGateCount is the size of the combined HITL gate: the number of tool
// calls still awaiting approval plus the number of change sets still awaiting
// review. The workflow waits while this is > 0 and re-invokes the runner once it
// reaches 0. Reading both sub-gates here (rather than pending_approvals alone)
// is what lets a turn blocked purely on file review wait for, and resume from,
// the same signal as tool approvals.
func UnresolvedGateCount(status *agentexecutionv1.AgentExecutionStatus) int {
	return len(status.GetPendingApprovals()) + CountAwaitingReview(status.GetFileChangeSets())
}

// GateResolved reports whether the combined HITL gate is fully clear — no
// pending approvals AND no change set awaiting review — so the turn may resume.
func GateResolved(status *agentexecutionv1.AgentExecutionStatus) bool {
	return UnresolvedGateCount(status) == 0
}

File: backend/services/stigmer-server/src/domain/agentexecution/filereview/gate.ts
/**
 * The unified HITL resume gate — ports filereview/gate.go. Two lifecycles
 * can block a single turn: tool approvals (pending_approvals) and file
 * review (file_change_sets awaiting a human decision); a turn is
 * unblocked only when BOTH are clear. These helpers are the one place
 * that truth is computed, so the workflow's wake condition,
 * SubmitApproval's signal, and SubmitFileDecision's signal can never
 * disagree about whether the turn may resume.
 */
import type { AgentExecutionStatus } from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/api_pb";
import { FileChangeSetStatus } from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/enum_pb";
import type { FileChangeSet } from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/filereview_pb";

/**
 * How many change sets still wait on a human decision (AWAITING_REVIEW).
 * A CAPTURING set is mid-turn (the runner has not returned yet) and a
 * DECIDED/RECONCILED/FAILED set no longer blocks the gate, so neither
 * counts.
 */
export function countAwaitingReview(changeSets: FileChangeSet[]): number {
  let n = 0;
  for (const cs of changeSets) {
    if (cs.status === FileChangeSetStatus.AWAITING_REVIEW) {
      n++;
    }
  }
  return n;
}

/** Whether the file-review sub-gate is clear. */
export function noChangeSetAwaitingReview(
  status: AgentExecutionStatus,
): boolean {
  return countAwaitingReview(status.fileChangeSets) === 0;
}

/**
 * The size of the combined HITL gate: tool calls still awaiting approval
 * plus change sets still awaiting review. The workflow waits while this
 * is > 0 and re-invokes the runner once it reaches 0.
 */
export function unresolvedGateCount(status: AgentExecutionStatus): number {
  return (
    status.pendingApprovals.length +
    countAwaitingReview(status.fileChangeSets)
  );
}

/** Whether the combined HITL gate is fully clear — the turn may resume. */
export function gateResolved(status: AgentExecutionStatus): boolean {
  return unresolvedGateCount(status) === 0;
}

/**
 * Whether some change set is DECIDED — verdicts recorded, runner
 * reconcile still owed. The workflow's wait loop uses this to tell a
 * LEGITIMATE empty gate from a broken one (a set decided before the gate
 * check — the DD-28 auto-keep or a fast human — leaves phase
 * WAITING_FOR_APPROVAL with a zero gate count; the correct response is an
 * immediate re-invoke to reconcile, never the empty-gate anomaly path).
 */
export function hasDecidedAwaitingReconcile(
  status: AgentExecutionStatus,
): boolean {
  return status.fileChangeSets.some(
    (cs) => cs.status === FileChangeSetStatus.DECIDED,
  );
}
